netaddr.go

```diff
@@ -163,58 +163,37 @@ func (err parseIPError) Error() string {
 }
 
 // parseIPv4 parses s as an IPv4 address (in form "192.168.0.1").
-func parseIPv4(in string) (ip IP, err error) {
-	s := in
-	var ip4 [4]byte
-
-	for i := 0; i < 4; i++ {
-		var (
-			j   int
-			acc uint16
-		)
-		// Parse one byte of digits. Bail if we overflow, stop at
-		// first non-digit.
-		//
-		// As of Go 1.15, don't try to factor this digit reading into
-		// a helper function. Its complexity is slightly too high for
-		// inlining, which ends up costing +50% in parse time.
-		for j = 0; j < len(s); j++ {
-			if s[j] < '0' || s[j] > '9' {
-				break
+func parseIPv4(s string) (ip IP, err error) {
+	var fields [3]uint8
+	var val, pos int
+	for i := 0; i < len(s); i++ {
+		if s[i] >= '0' && s[i] <= '9' {
+			val = val*10 + int(s[i]) - '0'
```

Comment:
When brackets matter:
```go
val = val*10 + int(s[i]) - '0'
```
  
```asm
  55010a:       48 8d 34 b6             lea    rsi,[rsi+rsi*4] ; rsi = rsi*4 + rsi -> val*5
  55010e:       48 8d 34 77             lea    rsi,[rdi+rsi*2] ; rsi*2 + rdi       -> val*10 + s[i]
  550112:       48 8d 76 d0             lea    rsi,[rsi-0x30]  ; rsi = rsi - 0x30  -> val*10 + s[i] - '0'
```

```go
val = val*10 + int(s[i] - '0')
```

```asm
  55010a:       48 8d 34 b6             lea    rsi,[rsi+rsi*4] ; rsi = rsi*4 + rsi      -> val*5
  55010e:       83 c7 d0                add    edi,0xffffffd0  ; edi = edi + 0xffffffd0 -> s[i] - '0'
  550111:       40 0f b6 ff             movzx  edi,dil         ;                        -> int(s[i] - '0')
  550115:       48 8d 34 77             lea    rsi,[rdi+rsi*2] ; rsi = rsi*2 + rdi      -> val*10 + int(s[i] - '0')
```

Reply:
Mind blowing. But probably worth leaving a comment that what you've written specifically compiles to less instructions. Or maybe there's room for a compiler optimization here @josharian? Beats me!

Reply:
Just an amusement I would not put a comment that we trick compiler here.
I would rather rely on benchmark numbers if this is optimized further or a regression is observed.

Reply:
Can't do this in the compiler, because they have different overflow/underflow semantics. Picking the right order of operations can definitely help.